### ISSUE TITLE
`pipe` loses generics (2)

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#35167]
**TypeScript Version:**  3.6.3

**Search Terms:**

**Code**

```ts
function pipeWith<A, B>(a: A, ab: (this: void, a: A) => B): B;

type Predicate<A> = (a: A) => boolean;

declare const filter: <A>(predicate: Predicate<A>) => (fa: A) => A;

declare function isString(candidate: any): candidate is string;

declare const r: 'foo';

const r2 = pipeWith(r, filter(isString)); // expected type "foo", actual any
const r3 = pipeWith(r, filter(v => isString(v))); // expected and actual type "foo"
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:** 

**Related Issues:** 

https://github.com/microsoft/TypeScript/issues/30727

**Comment by karol-majewski** [comment]
The missing signature:

```ts
declare function isString(candidate: any): candidate is string;
```

Full [TypeScript Playground](https://www.typescriptlang.org/play/?ssl=2&ssc=64&pln=2&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAcsCQB1LDACwB4BBAGngCEA+ACigC54H4oAjbmypYAztwBuOLMEZceASngBeFswXcmAbgBQoSLAQp02PPDEBlDDCyoA5mzBRUwGVAwhuzgJ4b4TlzcPc1F4UWtbO10dDG8SeAAFOFcnDzo1ZXgOblolVXh+HBwIEGdo-Wg4fzxwpCwIDxhudLYCZKxUz0T2zvS8tTZEeVyVNVpy8EqEMBqMeCb4AHJEIsXomdRamAAmFUJiMgpKNhhGRHrGtksI+wUFLXgAekf4EAAPEkwQYHhY+IAiFY4f5yTDIKAQPiobw6DZbADMeyIJHIVBOZwuIBgbAkoxCVhs9hxd3uTxe70+Hh+zmpYIhvziCEBRX+QA)

**Comment by OliverJAsh** [comment]
Thanks @karol-majewski, updated my description.

**Comment by jack-williams** [comment]
This is not really related to `pipe` specifically. For example:
```ts
const a: (x: "foo") => "foo" = filter(isString); // infers A as any
const b: (x: "foo") => "foo" = filter(v => isString(v)); // infers A as "foo"
```

The problem is that in the first case there is not floating argument lacking a type annotation, so any type inference is done by directly unifying `Predicate<A>` and `typeof isString`, which results in the unification of `A` with `any`.

In the second case the `v` parameter is free and subject to contextual typing which comes from the annotation in my example, or the first argument in your example. The contextual type of `v` is `"foo"` which then unifies `Predicate<A>` with `(v: "foo") => boolean`, and `A` with `"foo"`.

TLDR: eta expansion of a function introduces "wiggle" room that lets contextual typing make inferences. Without the exapansion you're committing to inferences obtained from the declared type of `isString`.

This is similar to why in the following case:
```ts
const a1 = isString;
const b1 = v => isString(v);
```
The type of `a1` is inferred correctly, but `b1` gets an implicit `any` error.

**Comment by OliverJAsh** [comment]
@jack-williams Thanks for your explanation! Should we label this as a suggestion?

**Comment by andrewbranch** [comment]
Worth noting that a kinda gross workaround [exists](https://www.typescriptlang.org/play/?ssl=11&ssc=42&pln=11&pc=77#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAcsCQB1LDACwB4BBAGngCEA+ACigC54H4oAjbmypYAztwBuOLMEZceASngBeFswXcmAbgBQOjAE8S8AApxgWMFAwg6a5fA7daS1fH44cEEFFS6doJCwCGB4ohhIWBA2MNx2bATmltYg3GYgFlY2dq5qbIjyLipqtP6B0HBIaJi4+GIAyhgwWKgA5tQAKuxWqBbAKdwd8AA+8GgA1qg4AO6oGvA9fSnwYvDhzW1l4BUhYRGx8ADkiJ6H-qGo4fAwAEwqhMRkFJRsMIyIUTFsDU0trQoKXQXK4wADM9yIJHIVFe70+IBgbAkxRWokaG1aSIBgKAA) 👀
